### PR TITLE
Fix: Crash on empty section layout when footer visibility is toggled

### PR DESCRIPTION
**Pull request by rebeccachin**
## Details

A crash occurs during layout when a section is updated to: 
1. be empty
2. hide the footer

Under these conditions, `initialLayoutAttributesForAppearingSupplementaryElement(ofKind elementKind: String, at elementIndexPath: IndexPath)`
is called with an **empty** IndexPath. This results in a crash when `elementIndexPath.section` is accessed. 

## Motivation and Context

Our project has a notion of a `collapsed` section, which is essentially an empty section. 
`Collapsed` sections do not show a footer view, but the expanded section may have a footer.

## How Has This Been Tested

I have modified the example project to reproduce the issue (see second commit on my PR). 
To test, I removed all of the items from a section that had a footer.
I observed that:
*without the fix*, the modified example crashes
*with the fix*, the modified example does not crash

I took at look at the test suite to see if I could add coverage for this test case, but I'm not sure what the your preferred strategy for testing this code path would be. The logic lives within the `MagazineLayout` class, which is not currently under test, as far as I can tell. 

**Note**: I'm not sure that the changes to example app are appropriate to bring back, but I wanted to be able to show you what was up. 

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project. ( I think so?) 
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by rebeccachin** [comment]
Oh boy, I'm not sure why travis is failing. It builds and tests pass on my local machine. 
I did notice that travis was picking up an old github username for me. Maybe it was unable to find my PR branch?

**Comment by bryankeller** [comment]
Thanks for investigating this @rebeccachin - I'm not able to reproduce the crash on my end though, even when making my code look like the second commit. Do you have a crash log?

When I remove the last item of a section that started with a couple items and a header and footer, it looks like `initialLayoutAttributesForAppearingSupplementaryElement` isn't even invoked.

If we did get into `initialLayoutAttributesForAppearingSupplementaryElement` with an `elementIndexPath` whose `isEmpty` property were `true`, I don't see where / why a crash would occur.

1.  `modelState.sectionIndicesToInsert.contains(elementIndexPath.section)` would fail, I assume?
2. `let movedSectionID = modelState.idForSectionModel(...` would return nil, I assume?
3. Finally, we're in the else block, which should just invoke super.

Obviously, some assumptions being made on my part since I can't actually repro 😛

What OS version / iOS simulator / crashing line? Even better would be a full stack trace.And if you have a video or GIF of the crash, that'd be great too! Thanks for your help on this.

**Comment by bryankeller** [comment]
@rebeccachin any info would be super helpful! Thanks!

**Comment by rebeccachin** [comment]
@bryankeller, Sorry for the long turn around here! 
I'm able to reproduce the crash on simulator at iOS 13.3, by removing all of the elements from the last section.

You're right `modelState.sectionIndicesToInsert.contains(elementIndexPath.section)` fails at `elementIndexPath.section`. 

```
(lldb) po elementIndexPath
0 elements

(lldb) po elementIndexPath.section
error: Execution was interrupted, reason: EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0).
The process has been returned to the state before expression evaluation.
(lldb) 
```

Here is the full stack trace of the crash: 

```
 * thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00007fff5202bdc2 libswiftUIKit.dylib`merged (extension in UIKit):Foundation.IndexPath.section.getter : Swift.Int + 34
    frame #1: 0x000000011025bb8c MagazineLayout`MagazineLayout.initialLayoutAttributesForAppearingSupplementaryElement(elementKind="MagazineLayoutSupplementaryViewKindSectionFooter", elementIndexPath=Foundation.IndexPath @ 0x00007ffedfd52e20, self=0x00007febf600ee60) at MagazineLayout.swift:537:68
    frame #2: 0x000000011025ce3d MagazineLayout`@objc MagazineLayout.initialLayoutAttributesForAppearingSupplementaryElement(ofKind:at:) at <compiler-generated>:0
    frame #3: 0x00007fff4787397d UIKitCore`-[UICollectionView _doubleSidedAnimationsForView:withStartingLayoutAttributes:startingLayout:endingLayoutAttributes:endingLayout:withAnimationSetup:animationCompletion:enableCustomAnimations:customAnimationsType:] + 745
    frame #4: 0x00007fff4788a433 UIKitCore`-[UICollectionView _viewAnimationsForCurrentUpdate] + 13003
    frame #5: 0x00007fff4788f1a6 UIKitCore`__71-[UICollectionView _updateWithItems:tentativelyForReordering:animator:]_block_invoke.1877 + 118
    frame #6: 0x00007fff4856a626 UIKitCore`+[UIView(Animation) performWithoutAnimation:] + 84
    frame #7: 0x00007fff4788dfdb UIKitCore`-[UICollectionView _updateWithItems:tentativelyForReordering:animator:] + 4002
    frame #8: 0x00007fff47886a8e UIKitCore`-[UICollectionView _endItemAnimationsWithInvalidationContext:tentativelyForReordering:animator:] + 17015
    frame #9: 0x00007fff478905ae UIKitCore`-[UICollectionView _endUpdatesWithInvalidationContext:tentativelyForReordering:animator:] + 71
    frame #10: 0x00007fff47890906 UIKitCore`-[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:animator:] + 455
    frame #11: 0x00007fff4789071c UIKitCore`-[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:] + 90
    frame #12: 0x00007fff4789069f UIKitCore`-[UICollectionView _performBatchUpdates:completion:invalidationContext:] + 74
    frame #13: 0x00007fff478905f4 UIKitCore`-[UICollectionView performBatchUpdates:completion:] + 53
  * frame #14: 0x000000010fec3c01 MagazineLayoutExample`ViewController.collectionView(collectionView=0x00007febf8825600, indexPath=Foundation.IndexPath @ 0x00007ffedfd54dd0, self=0x00007febf66077e0) at ViewController.swift:405:20
    frame #15: 0x000000010fec3ff7 MagazineLayoutExample`@objc ViewController.collectionView(_:didSelectItemAt:) at <compiler-generated>:0
    frame #16: 0x00007fff478652c4 UIKitCore`-[UICollectionView _selectItemAtIndexPath:animated:scrollPosition:notifyDelegate:deselectPrevious:] + 633
    frame #17: 0x00007fff47892678 UIKitCore`-[UICollectionView _userSelectItemAtIndexPath:] + 197
    frame #18: 0x00007fff478928ce UIKitCore`-[UICollectionView touchesEnded:withEvent:] + 583
    frame #19: 0x00007fff480bf863 UIKitCore`forwardTouchMethod + 340
    frame #20: 0x00007fff480bf932 UIKitCore`-[UIResponder touchesEnded:withEvent:] + 49
    frame #21: 0x00007fff480bf863 UIKitCore`forwardTouchMethod + 340
    frame #22: 0x00007fff480bf932 UIKitCore`-[UIResponder touchesEnded:withEvent:] + 49
    frame #23: 0x00007fff47c33b82 UIKitCore`_UIGestureEnvironmentUpdate + 10042
    frame #24: 0x00007fff47c3140a UIKitCore`-[UIGestureEnvironment _deliverEvent:toGestureRecognizers:usingBlock:] + 467
    frame #25: 0x00007fff47c3117f UIKitCore`-[UIGestureEnvironment _updateForEvent:window:] + 200
    frame #26: 0x00007fff480d04b0 UIKitCore`-[UIWindow sendEvent:] + 4574
    frame #27: 0x00007fff480ab53b UIKitCore`-[UIApplication sendEvent:] + 356
    frame #28: 0x00007fff4812c71a UIKitCore`__dispatchPreprocessedEventFromEventQueue + 6847
    frame #29: 0x00007fff4812f1e0 UIKitCore`__handleEventQueueInternal + 5980
    frame #30: 0x00007fff23bd4471 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
    frame #31: 0x00007fff23bd439c CoreFoundation`__CFRunLoopDoSource0 + 76
    frame #32: 0x00007fff23bd3b74 CoreFoundation`__CFRunLoopDoSources0 + 180
    frame #33: 0x00007fff23bce87f CoreFoundation`__CFRunLoopRun + 1263
    frame #34: 0x00007fff23bce066 CoreFoundation`CFRunLoopRunSpecific + 438
    frame #35: 0x00007fff384c0bb0 GraphicsServices`GSEventRunModal + 65
    frame #36: 0x00007fff48092d4d UIKitCore`UIApplicationMain + 1621
    frame #37: 0x000000010fec7e08 MagazineLayoutExample`main at AppDelegate.swift:19:13
    frame #38: 0x00007fff5227ec25 libdyld.dylib`start + 1
    frame #39: 0x00007fff5227ec25 libdyld.dylib`start + 1
```

File: MagazineLayout/Public/MagazineLayout.swift [hunk 1]
```diff
@@ -558,6 +558,12 @@ public final class MagazineLayout: UICollectionViewLayout {
     at elementIndexPath: IndexPath)
     -> UICollectionViewLayoutAttributes?
   {
+    if elementIndexPath.isEmpty {
+      return super.initialLayoutAttributesForAppearingSupplementaryElement(
+        ofKind: elementKind,
+        at: elementIndexPath)
+    }
```
Comment (bryankeller) [inline MagazineLayout/Public/MagazineLayout.swift]:
```suggestion
   // If a supplementary view's visibility changes to `.hidden` due to a data source change, this
    // function will get invoked with an `elementIndexPath` that crashes when its `section` is
    // accessed.
    guard !elementIndexPath.isEmpty else {
      return super.initialLayoutAttributesForAppearingSupplementaryElement(
        ofKind: elementKind,
        at: elementIndexPath)
    }
```

File: Example/MagazineLayoutExample/ViewController.swift [hunk 2]
```diff
@@ -403,13 +403,8 @@ extension ViewController: UICollectionViewDelegate {
...
     collectionView.performBatchUpdates({
-      if dataSource.numberOfItemsInSection(withIndex: indexPath.section) > 1 {
-        dataSource.removeItem(atItemIndex: indexPath.item, inSectionAtIndex: indexPath.section)
-        collectionView.deleteItems(at: [indexPath])
-      } else {
-        dataSource.removeSection(atSectionIndex: indexPath.section)
-        collectionView.deleteSections(IndexSet(integer: indexPath.section))
-      }
+      dataSource.removeItem(atItemIndex: indexPath.item, inSectionAtIndex: indexPath.section)
+      collectionView.deleteItems(at: [indexPath])
```
Comment (bryankeller) [inline Example/MagazineLayoutExample/ViewController.swift]: Same here - let's remove then we can merge

**Comment by bryankeller** [comment]
Thanks a bunch @rebeccachin - this fix looks solid. I'll have a separate PR that fixes the wonky animation, but this at least fixes the crash.

**Comment by rebeccachin** [comment]
@bryankeller Thanks! I'm sorry I completely dropped this.

**Comment by bryankeller** [comment]
No worries at all @rebeccachin - my feedback was super minor anyways 🙂

**Pull request merged**